lib/karto_sdk/include/karto_sdk/M3RScanMatcher.h

```diff
@@ -0,0 +1,122 @@
+#ifndef M3RSM__H
+#define M3RSM__H
+
+#include <stdint.h>
+#include <map>
+#include <vector>
+#include <unordered_map>
+#include <queue>
+#include <algorithm>
+#include <chrono>
+#include <utility>
+#include <string>
+
+#include "Eigen/Core"
+#include "rclcpp/rclcpp.hpp"
+#include "karto_sdk/Karto.h"
+#include "karto_sdk/Mapper.h"
+
+namespace {
+  typedef std::vector<std::vector<uint8_t>> DataMatrix;
+}
+
+namespace karto
+{
+
+class M3RSM_scan;
+
+// TODO: Implement this class.
+class LookupTable {
+
+  public:
+    // Builds highest resolution lookup table from base scans
+    LookupTable(const std::vector<LocalizedRangeScan>& base_scans, int decimation_factor, int kernel_width);
+    // Copy constructor
+    LookupTable(const LookupTable& table);
+    // Builds lower resolution lookup table from higher resolution lookup table
+    static LookupTable GetLowerResTable(const LookupTable& higher_res_table);
+    double computeCost(const LocalizedRangeScan& scan);
+    bool isMaxResolution();
+    int getTableSize();
+
+  private:
+    DataMatrix data_;
+    bool is_max_res_;
+    int decimation_factor_;
+    int kernel_width_;
+
+};  // class LookupTable
+
+class MockLookupTable {
+
+  public:
+    MockLookupTable(const DataMatrix& data, int decimation_factor, int kernel_width);
+    // Builds lower resolution lookup table from higher resolution lookup table
+    MockLookupTable(const MockLookupTable& table);
+    static MockLookupTable GetLowerResTable(const MockLookupTable& higher_res_table);
+    double computeCost(const std::vector<std::pair<int, int>>& occupied_cells);
+    bool isMaxResolution();
+    int getTableSize();
+    DataMatrix data_;
+    bool is_max_res_;
+    int decimation_factor_;
+    int kernel_width_;
+};  // class MockLookupTable
+
+class LookupTableManager {
```

Comment:
@LucasJSch meta: only to not lose track of our discussion today.

Rasterized image pyramids can be cached (or _managed_) for a single scan matching request, but it is not clear how and if we can cache these pyramids **across** scan matching requests. We could simply persist the cache but it is unlikely we'll be given the same set of reference scans often enough for the cache to be worth it. We could cache a pyramid per scan and merge them, but the candidate scan pose (and thus the search window pose) changes over time. It is hard to tell whether we can freely apply translations and rotations to decimated rasterized images or not. 

How can we maximize cache hits across scan matching runs? I'll sleep on it.